### ISSUE TITLE
Order not maintained in Helm toYaml function when using multiple alphanumeric nested keys

### ISSUE THREAD

**Issue by JacobAmar** [helm/helm#12537]
Encountered an unusual behavior with Helm's toYaml function when it's deployed to generate a configmap file. In this case, the toYaml function is used to generate, directly into the configmap template, a YAML file from a configmap.

It appears that the problem arises when seven or more nested alphanumeric keys are used. The order isn't retained consistently when rendering. Interestingly, this doesn't seem to happen with alphabetic keys.

Below is a step-by-step reproduction of the issue, and its effects can be clearly seen by running the provided diff command multiple times. This inconsistency barometer could cause argoCD applications to lose synchronization.

Template used: configmap.yaml

```yaml
apiVersion: v1
kind: List
metadata:
  name: "configmap-list"
items:
- kind: ConfigMap
  apiVersion: v1
  metadata:
    name: "test"
  binaryData: 
    {{- range $key, $value := .Values.configmaps.kv_data }}
    {{ $key }}: >-
       {{ toYaml $value | trim | b64enc }}
    {{- end }}
```
Values used:
```yaml
  kv_data:
    config.yaml:
      some:
        test:
          nested:
            loop:
              98c8ef5d: "200"
              a0886a2d: "501"
              a0a733f5: "501"
              aa34bb05: 10
              aa32a1s0: 2
              aa32a120: 2
              aa323asf: 5
```

Diff command:  `helm template . | yq e '.items[0].binaryData."config.yaml"' | base64 -d  > 01config.yaml && helm template . | yq e '.items[0].binaryData."config.yaml"' | base64 -d  > 02config.yaml  && diff -u 01config.yaml 02config.yaml`

if you will run the cummand multiple times you will see that helm from time to time change items order from the `loop` map

Oddly, if there are fewer than six keys, the problem doesn't occur, but from six keys and up, the issue reappears.

My temporary workaround is to utilize a list within the loop keys instead:

```yaml
configmaps:
  kv_data:
    config.yaml:
      some:
        test:
          nested:
            loop:
              - 98c8ef5d: "200"
              - a0886a2d: "501"
              - a0a733f5: "501"
              - aa34bb05: 10
              - aa32a1s0: 2
              - aa32a120: 2
              - aa323asf: 5
```

However, this workaround may not be sustainable in the long run as I frequently use multiple value files and merge loop maps with other loop maps for overwriting purposes.

Order inconsistency with multiple alphanumeric nested keys in Helm's toYaml, causing argoCD applications to lose sync

Output of `helm version`:
```
version.BuildInfo{Version:"v3.13.1", GitCommit:"3547a4b5bf5edb5478ce352e18858d8a552a4110", GitTreeState:"clean", GoVersion:"go1.21.3"}
```
Output of `kubectl version`:
```
Client Version: v1.28.0
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.24.17-eks-f8587cb
```
Cloud Provider/Platform (AKS, GKE, Minikube etc.):  EKS

**Comment by noaramon** [comment]
+1

**Comment by gjenkins8** [comment]
Helm [parses](https://github.com/helm/helm/blob/main/pkg/chartutil/values.go#L108) yaml values into go maps. Which have no guarantee of ordering.  In fact, https://go.dev/blog/maps ("Iteration Order") states that the order might change every time you iterate! So unless I'm mistaken here, this has never been guaranteed. And might have worked sometimes by change/implementation detail.

 If you think there was an ordering guarantee in a previous version of Helm (which is now broken), we can try and triage.
 
If not, and stable ordering is desired as a feature. Helm uses "sigs.k8s.io/yaml" to parse the values inputs, so ultimately any changes to support stable map ordering would need to be supported by this library (or equivalent).

**Comment by mot256** [comment]
We ran into this same issue. We use sha256sum over the output of toYaml to get a hash for a secret name. Due to this issue we get random hash values for the same helm values. 
We found toJson and toPrettyJson functions are deterministic (probably because their output has json properties ordered alphabetically). If we could have an option for toYaml that also sorts properties by name, we could have a deterministic output.
